crates/ruma-common/src/events/poll/start/content_serde.rs

```diff
@@ -0,0 +1,217 @@
+            automated,
+        } = value;
+
+        Self {
+            poll: poll_start.into(),
+            text: message.into(),
+            #[cfg(feature = "unstable-msc3955")]
+            automated,
+        }
+    }
+}
+
+impl From<PollStartEventContent> for PollStartEventContentUnstableSerDeHelper {
+    fn from(value: PollStartEventContent) -> Self {
+        let PollStartEventContent {
+            poll,
+            text,
+            #[cfg(feature = "unstable-msc3955")]
+            automated,
+        } = value;
+
+        Self {
+            poll_start: poll.into(),
+            message: text.into(),
+            #[cfg(feature = "unstable-msc3955")]
+            automated,
+        }
+    }
+}
+
+/// A block for poll content.
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+struct PollStartContentBlockUnstableSerDeHelper {
+    question: PollQuestionUnstableSerDeHelper,
+
+    #[serde(default)]
+    kind: PollKind,
+
+    #[serde(
+        default = "PollContentBlock::default_max_selections",
+        skip_serializing_if = "PollContentBlock::max_selections_is_default"
+    )]
+    max_selections: UInt,
+
+    answers: PollAnswersUnstableSerDeHelper,
+}
+
+impl From<PollStartContentBlockUnstableSerDeHelper> for PollContentBlock {
+    fn from(value: PollStartContentBlockUnstableSerDeHelper) -> Self {
+        let PollStartContentBlockUnstableSerDeHelper { question, kind, max_selections, answers } =
+            value;
+
+        Self { question: question.into(), kind, max_selections, answers: answers.into() }
+    }
+}
+
+impl From<PollContentBlock> for PollStartContentBlockUnstableSerDeHelper {
+    fn from(value: PollContentBlock) -> Self {
+        let PollContentBlock { question, kind, max_selections, answers } = value;
+
+        Self { question: question.into(), kind, max_selections, answers: answers.into() }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct PollQuestionUnstableSerDeHelper {
+    #[serde(flatten)]
+    message: MessageContentBlock,
+}
+
+impl From<PollQuestionUnstableSerDeHelper> for PollQuestion {
+    fn from(value: PollQuestionUnstableSerDeHelper) -> Self {
+        Self { text: value.message.into() }
+    }
+}
+
+impl From<PollQuestion> for PollQuestionUnstableSerDeHelper {
+    fn from(value: PollQuestion) -> Self {
+        Self { message: value.text.into() }
+    }
+}
+
+#[derive(Serialize)]
+struct PollAnswersUnstableSerDeHelper(Vec<PollAnswerUnstableSerDeHelper>);
+
+impl<'de> Deserialize<'de> for PollAnswersUnstableSerDeHelper {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let mut answers = Vec::<PollAnswerUnstableSerDeHelper>::deserialize(deserializer)?;
+
+        if answers.is_empty() {
+            return Err(de::Error::invalid_length(0, &"at least 1 answer"));
+        }
+
+        answers.truncate(20);
```

Comment:
Should the deserializer be modifying the data like this? I'd expect a deserializer to simply transform the data as-is and leave the domain logic of checking/truncating the answers to someone else

Reply:
I've been asked to comment on this. The MSC states:

> * `answers` - Array of options users can select. Each entry is an object with an `m.text` content block, similar to `question`, and an opaque string field `m.id` for use in response events. More
  blocks might be added in the future. Clients should treat each entry similar to how they would an
  `m.message` event. The array is truncated to 20 maximum options.

Since there are no strings attached or anything, i.e. the array is *always* truncated, I think this is right. Maybe we should check if the truncation does anything and log it in that case though.

Reply:
Sorry, just to clarify, I think it's good that Ruma does the truncation but wasn't convinced that it should be done by the deserializer. But I think it's a fair argument that the deserializer can reasonably expect (and coerce) spec compliant data just as it expects valid JSON.

It sounds like this file may need to be removed anyway (https://github.com/ruma/ruma/pull/1589#discussion_r1275108013)

Reply:
Note that even if the file is gon, the truncation still happens in `(unstable_)poll_answers_serde.rs`